File: python/rpn-calculator.py
from math import factorial
def readNumber(string: str,index: int):
    number = ""
    while index < len(string) and isNumber(string[index]):
        number += string[index]
        index += 1
    return float(number), index


def isOperator(token):
    operators = ['+', '-', '*', '/', '%','^']
    return token in operators

def isNumber(token):
    return (token >= '0' and token <= '9') or token == "."



def toRpn(string: str):
    """
    Converts infix notation to reverse polish notation
    """
    precedence = {
        '(': 0,
        '-': 1,
        '+': 1,
        '*': 2,
        '/': 2,
        '%': 2,
        '^': 3,
    }

    i = 0
    fin = []
    ops = []

    while i < len(string):
        token = string[i]
        if isNumber(token):
            number, i = readNumber(string,i)
            fin.append(number)
            continue
        if isOperator(token):
            top = ops[-1] if ops else None
            if top is not None and precedence[top] >= precedence[token]:
                fin.append(ops.pop())
            ops.append(token)
            i += 1
            continue
        if token == '(':
            ops.append(token)
            i += 1
            continue
        if token == ')':
            while True:
                operator = ops.pop()
                if operator == '(':
                    break
                fin.append(operator)
                if not ops:
                    break
            i += 1
            continue
        i += 1
    while ops:
        fin.append(ops.pop())
    return fin

def calculate_rpn(rpn: list):
    """
    Calculates the result of an expression in reverse polish notation
    """
    stack = []
    for token in rpn:
        if isOperator(token):
            a = stack.pop()
            b = stack.pop()
            if token == '+':
                stack.append(b + a)
            elif token == '-':
                stack.append(b - a)
            elif token == '*':
                stack.append(b * a)
            elif token == '/':
                stack.append(b / a)
            elif token == '%':
                stack.append(b % a)
            elif token == '^':
                stack.append(b ** a)
        else:
            stack.append(token)
    return stack[0]
        

print ("90165: ", calculate_rpn(toRpn("27+38+81+48*33*53+91*53+82*14+96")))
print ("616222: ", calculate_rpn(toRpn("22*26*53+66*8+7*76*25*44+78+100")))
print (calculate_rpn(toRpn("(22+4)*4")))
print (calculate_rpn(toRpn("3^2+2")))

File: cpp/rpn-calculator.cpp
#include <string>
#include <stack>
#include <iostream>
#include <math.h>
using namespace std;

struct element
{
    string value;
    struct element *next;
};

void push(struct element **anchor, string value)
{
    struct element *new_element = new struct element;
    new_element->value = value;
    new_element->next = nullptr;
    if (*anchor == nullptr)
    {
        *anchor = new_element;
    }
    else
    {
        struct element *current = *anchor;
        while (current->next != nullptr)
        {
            current = current->next;
        }
        current->next = new_element;
    }
}

struct element *pop(struct element **anchor)
{
    struct element *current = *anchor;
    if (current == nullptr)
    {
        return nullptr;
    }else{
        // TODO: Fehler wenn nur ein Element in der Liste ist
        while(current->next->next != nullptr)
        {
            current = current->next;
        }
        struct element *temp = current->next;
        current->next = nullptr;
        return temp;
    }
}

struct element *top(struct element **head)
{
    struct element *current = *head;
    if (current == nullptr)
    {
        return nullptr;
    }
    else
    {
        while(current->next != nullptr)
        {
            current = current->next;
        }
        return current;
    }
}

bool is_operator(char c)
{
    return (c == '+' || c == '-' || c == '*' || c == '/' || c == '^');
}

bool is_number(char c)
{
    return (c >= '0' && c <= '9') || c == '.';
}

string read_number(string &s, int i)
{
    string number = "";
    while (i < s.length() && is_number(s[i]))
    {
        number += s[i];
        i++;
    }
    return number;
}

int get_precedence(char c)
{
    if (c == '+' || c == '-')
        return 1;
    if (c == '*' || c == '/' || c == '%')
        return 2;
    if (c == '^')
        return 3;
    return 0;
}

stack<string> to_rpn(string &s){
    int i = 0, l = s.length();
    struct element *fin = nullptr;
    struct element *ops = nullptr;

    while (i < l){
        char token = s[i];

        if (is_number(token))
        {
            string number = read_number(s, i);
            push(&fin, number);
            i += number.length();
            continue;
        }
        if (is_operator(token))
        {
            if (ops != nullptr && get_precedence(token) <= get_precedence(ops.top()[0]))
            {
                
                push(&fin,top(&ops)->value);
                pop(&ops);
            }
            push(&ops,string(1, token));
            i++;
            continue;
        }
        if (token == '(')
        {
            push(&ops,string(1, token));
            i++;
            continue;
        }
        if (token == ')')
        {
            do {
                string op = top(&ops)->value;
                pop(&ops);
                if (op[0] == ')')
                    break;
                push(&fin,op);
            }while( ops != nullptr);
            i++;
            continue;
        }
        i++;
    }
    while (ops != nullptr)
    {
        fin.push(ops.top());
        ops.pop();
    }
    return fin;
}

string get_value_of_expression(string b, string a, char op)
{
    if (op == '+')
        return to_string(stod(b) + stod(a));
    if (op == '-')
        return to_string(stod(b) - stod(a));
    if (op == '*')
        return to_string(stod(b) * stod(a));
    if (op == '/')
        return to_string(stod(b) / stod(a));
    if (op == '%')
        return to_string(stoi(b) % stoi(a));
    if (op == '^')
        return to_string(pow(stod(b), stod(a)));
    return "";
}

string calculate_rpn(stack<string> &rpn)
{
    stack<string> values;
    while (!rpn.empty())
    {
        string token = rpn.top();
        rpn.pop();
        if (is_operator(token[0]))
        {
            string a = values.top();
            values.pop();
            string b = values.top();
            values.pop();
            values.push(get_value_of_expression(b, a, token[0]));
        }
        else
        {
            values.push(token);
        }
    }
    return values.top();
}

int main()
{
    cout << "test" << endl;
    string s = "2+3";
    stack<string> rpn = to_rpn(s);
    cout << calculate_rpn(rpn);
    return 0;
}